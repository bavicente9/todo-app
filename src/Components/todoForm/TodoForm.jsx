
import { useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { selectTheme } from '../../features/theme/themeSlice'
import {addTodo} from '../../features/todosList/todosSlice'
import './TodoForm.scss'

const TodoForm = () => {
    const [textValue, setTextValue] = useState('')
    const [checkboxValue,setCheckboxValue ] = useState(false)

    const theme = useSelector(selectTheme)
    const dispatch = useDispatch()
    

    const handleCheckbox = (e) => {
        e.preventDefault()
        setCheckboxValue(e.target.checked)
    }

    const handleOnChangeText = (e) => {
        e.preventDefault()
        setTextValue(e.target.value)

    }
    
    const onSubmit = (e) => {
        e.preventDefault()
        const id = new Date().valueOf()

        //create the todo item
        const todoItem =  {
            id,
            text: textValue,
            active: !checkboxValue
        }
        //save the item into the store
        dispatch(addTodo(todoItem))
        
        //clean form
        setTextValue('')
    }

    return (
        <div className='todoForm'>
            <form
                className={`form_container form_container--${theme}`}
                onSubmit={onSubmit} >

                <input
                    name='Todo completed state'
                    className="input_checkBox "
                    type='checkbox'
                    onChange = {e => handleCheckbox(e)}
                />
                <input
                    name='todo text'
                    className={`input_text input_text--${theme}`}
                    required={true}
                    type='text' placeholder="Create a new todo..."
                    onChange={e => handleOnChangeText(e)}
                    value={textValue} />

            </form>
        </div>
    )
}


export default TodoForm;