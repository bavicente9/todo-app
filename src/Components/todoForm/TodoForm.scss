@import '../../Utils/style/globalVariables.scss';
@import '../../Utils/style/checkboxTodo.scss';


.todoForm {
    .form_container {
        padding: 15px 0;
        margin: 15px auto;
        
        min-width: $globalContainer-minWidth;
        max-width: $globalContainer-maxWidth;
        
        display: flex;
        .input_checkBox{
            
        }
        .input_text {
            background: none;
            border: none;
            outline: none;

            width: calc(90% - 33px);
            margin-left: 10px;
            color: $veryDarkGrayishBlue;
        }

        .input_text--dark {
            color: $veryLightGrayishBlue;
        }

        .input_text--light {
            color: $veryDarkGrayishBlue;
        }
    }

    .form_container--dark {
        background-color: $veryDarkDesaturatedBlue;
    }

    .form_container--light {
        background-color: $veryLightGray;
    }

    @media (max-width:$globalMedia-maxWidth) {

        .form_container {
            min-width: $globalMobileContainer-minWidth;
            max-width: $globalMobileContainer-maxWidth;

        }
    }
}