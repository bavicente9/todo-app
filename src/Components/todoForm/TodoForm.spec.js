
import renderer from 'react-test-renderer';
import TodoForm from "./TodoForm";
import componentWithMockStore from "../../Utils/test/componentWithMockStore";



describe('todo input', () => {
    let store
    let component;
    beforeEach(() => {
         //render a component with a mock Store
        const initialStore = {
            theme: {value: 'dark'},
            todos: {
                entities: [
                    {
                        id: '1',
                        text: 'text example',
                        active: true
                    },
                    {
                        id: '2',
                        text: 'text example number 2',
                        active: false
                    },
                ],
                showList: 'all',
                counterActive: 2
            }
        }

        const { component: com, store: sto } = componentWithMockStore(<TodoForm />, initialStore)
        store = sto
        component = com
    });

    it('match with snapshot', () => {
        expect(component.toJSON()).toMatchSnapshot();
    })



    it('call a dispatch when submit the box', () => {

        renderer.act(() => {
            const ev = {
                preventDefault: jest.fn()
            }
            component.root.findByType('form').props.onSubmit(ev);
        })

        expect(store.dispatch).toHaveBeenCalledTimes(1);

    })

    it('clean the input then of  the dispatch call', () => {
        let input;

        renderer.act(() => {

            const ev = {
                preventDefault: jest.fn(),
                target: {
                    value: 'Example'
                }
            }

            input = component.root.findByProps({ name: 'todo text' })
            input.props.onChange(ev)
            component.root.findByType('form').props.onSubmit(ev);
        })

        expect(input.props.value).toBe('')
    })


})