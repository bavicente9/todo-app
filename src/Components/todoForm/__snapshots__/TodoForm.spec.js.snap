// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todo input match with snapshot 1`] = `
<div
  className="todoForm"
>
  <form
    className="form_container form_container--dark"
    onSubmit={[Function]}
  >
    <input
      className="input_checkBox "
      maxlength="280"
      name="Todo completed state"
      onChange={[Function]}
      type="checkbox"
    />
    <input
      className="input_text input_text--dark"
      name="todo text"
      onChange={[Function]}
      placeholder="Create a new todo..."
      required={true}
      type="text"
      value=""
    />
  </form>
</div>
`;
